### ISSUE TITLE
sign for any positive timedelta should be/compare equal

### ISSUE THREAD

**Issue by gerritholl** [numpy/numpy#8463]
The current behaviour of `sign` on `timedelta64` objects is somewhat surprising:    

```
In [64]: sign(np.timedelta64(3, 'm')) == sign(np.timedelta64(3, 's'))
Out[64]: False
```

I'm not sure if the return type for the sign of a `timedelta64` should even be a `timedelta64` or or simply numerical.  Of course, normally `sign(x)` returns a result of the same `dtype` as `x`, but for any numerical data we have `uint8(1) == int8(1) == float32(1) == ....` etc.  I'm not sure what the best return value would be for the sign of a `timedelta64`, but I think it would be desirable if their signs compared equal.

**Comment by shoyer** [comment]
It is very strange that `sign` of `timedelta64` returns another `timedelta64`:
```
In [29]: np.sign(np.timedelta64(3, 'm'))
Out[29]: numpy.timedelta64(1,'m')

In [30]: np.sign(np.timedelta64(3, 's'))
Out[30]: numpy.timedelta64(1,'s')
```

I think we should make these return integers.

**Comment by seberg** [comment]
I would agree, might be nothing more then changing the ufunc signature necessary.

**Comment by mhvk** [comment]
Note that `np.sign` does generally return an object of the same type (even complex for complex, though it only looks at the sign of the real component). But I agree that here returning an integer is more logical: by multiplying with it you get an absolute value.

p.s. I hadn't quite realised that in many ways `timedelta64` is like a `Quantity` -- it even has a unit -- though it is not a very well-behaved one by storing its value as an integer: 
```
np.timedelta64(1, 'm') / 60 == np.timedelta64(1, 's')
# False
np.timedelta64(1, 'm') / 60 == np.timedelta64(0, 's')
# True
u.Quantity(1, 'min') / 60. == u.Quantity(1., 's')
# array(True, dtype=bool)
```

**Comment by charris** [comment]
@mhvk That reminds me that I've been toying with the idea of adopting the astropy double-double implementation of time. I'd really like a quad precision float or maybe int128, but double-double might be a more accessible type.

**Comment by mhvk** [comment]
@charris - for numpy, I think the largest benefit would come from having an actual quad-precision float, rather than from "faking" it with a double-double (for `Time`, there was an advantage of the double-double, in that the "Standard of Fundamenal Astronomy" library splits time in day + fraction). cc @aarchiba, who  mentioned (vague) plans for this with not too long ago (mostly in the context of simulations of few-body systems at the level of accuracy necessary for pulsar timing).

**Comment by WarrenWeckesser** [comment]
Cross reference, for readers of this issue: https://github.com/numpy/numpy/issues/29485

The behavior requested in that issue (that `np.sign(-0.0)` should return `-0.0`) demonstrates why `sign`  should not necessarily return integers.  Returning the same type as the input also allows `sign` to return `nan` when the input is `nan`.

**Comment by HansBrende** [comment]
@WarrenWeckesser after thinking a lot about this case since you linked it to the one I raised, I personally believe that there are only 2 logical implementations for np.sign(timedelta).

Implementation 1:

```python
raise TypeError
```

Implementation 2:

```python
if x > 0:
    return np.timedelta64(1)
if x < 0:
    return np.timedelta64(-1)
return np.timedelta64(0)
```
I.e., the output would be still along the "time axis" (similar to how the sign of a complex number preserves angle, so it is important that the output is still in the "time" domain), however, the specific time _unit of measure_ would be discarded in favor of "generic time units". I think this would be the most mathematically accurate result if numpy chooses to give _any_ result other than an error.

**Comment by seberg** [comment]
The complex/same type argument doesn't seem right to me, the correct thing (with the exception of `NaT`) seems to me to be to return an integer.
Marten mentioned that this means (for real value) `x * sign(x) == abs(x)` for example.  And if you think of the generalization for complex `x / abs(x)` also would be a unitless value.

A timedelta return, breaks almost all of the ["basic properties"](https://en.wikipedia.org/wiki/Sign_function#Basic_properties).

**Comment by HansBrende** [comment]
@seberg absolute values are, mathematically speaking, not defined on units of measure at all (i.e. |m| is meaningless), but given numpy's current implementation of that operation for timedelta (which is: "|scalar*uom| = |scalar|*uom"), I think you're right: the sign method should probably return a value in the domain of the wrapped scalar (ℤ in this case).

(Only question I'd have would be... which "flavor" of ℤ? The built-in `int`? `int64` to match `timedelta64`? `int8` for the minimum size sufficient? Technically even `timedelta64` meets this requirement since it is masquerading as ℤ via its superclass `numbers.Integral`. And don't forget about `TypeError`, still a plausible candidate in my opinion, which also meets this requirement since `typing.Never` ⊂ ℤ 😆.

You also mentioned `NaT` which poses another wrinkle, also implying that `timedelta64` is definitely not in ℤ. Shouldn't the output be floating point then, unless you resign yourself to mixed-domain outputs?)

**Comment by WarrenWeckesser** [comment]
If anyone is interested, I'd like to revive and resolve this issue.  Specifically, the issue is the return type and value of `sign(timedelta64)`.

I'm now leaning towards returning a `float64`, with the possible return values being -1.0, 0.0, 1.0 and `nan`.  Then, for a `timedelta64` instance `t`, the decomposition `t = sign(t)*abs(t)` works. (Currently it does not, because `sign(t)` will be a `timedelta64` instance, and multiplication is not defined for two `timedelta64` instances.)

@seberg, is the floating point result the option you preferred?

If we want to make this change, what would be the best way to make the transition?

**Comment by mhvk** [comment]
Agreed with `float`, mostly because one can then use `NaN` for the sign of any `NaT` -- something not possible with `int`.

**Comment by seberg** [comment]
> @seberg, is the floating point result the option you preferred?

I am happy with either float or integer+error but yeah, float for NaN support may be slightly better since we have no perfect result type.

> If we want to make this change, what would be the best way to make the transition?

Should we do a `FutureWarning`?  I am always a bit unsure about `FutureWarning`s when there is no good way to avoid them except by filtering them out explicitly.  (I.e. it feels like the `FutureWarning` may make more trouble than just flipping the switch here.)

Let's start on the mailing list? Maybe you have a better idea than me on how to approach this. I half suspect that there shouldn't really be any code that notices the change or code that might silently do the wrong thing if we do it, so maybe warnings are more noise than the actual change, but...
I half wonder if even a `DeprecationWarning` could make sense even if wrong, just because that is hidden by default (i.e. _if_ a library is affected, it get's time to make sure that is OK).

Anyway, I doube this is worth bringing out big machinery (like opt-in warnings and flip the switch in a 3.0 only), but maybe others disagree.

**Comment by WarrenWeckesser** [comment]
> Maybe you have a better idea than me on how to approach this. ...

No, I don't have anything that I consider a *good* idea.

`sign` is a ufunc, and the change we are talking about is changing the `timedelta64` type signature of `sign` from `m->m` to `m->d`.

I don't recall if there is already a standard way to make such a change that follows our preferred deprecation cycle.  How does a ufunc generate a warning when given a `timedelta64` argument?  How do we provide a mechanism for a user to opt in to the new behavior (and eliminate the warning) during the deprecation period? (And, preferably, opting in should not require *another* change when the deprecation period ends.)

I guess the simplest option is to just change it, with no warnings and no deprecation.  Call it a bug fix of a very old bug.

**Comment by WarrenWeckesser** [comment]
I opened https://github.com/numpy/numpy/pull/30194 (draft status) to see how involved the change with no warnings or deprecation would be.

**Comment by seberg** [comment]
> I guess the simplest option is to just change it, with no warnings and no deprecation. Call it a bug fix of a very old bug.

Yeah, which I am fine with, I just don't believe in the "bug fix" as more than a small part of the argument.  To me, the main argument has to be:
* We can't really see it leading to bad results for users.
* Only few people should notice the change (even among users).
* Generally, the warning is more annoying than any transition path.

> How does a ufunc generate a warning when given a timedelta64 argument? 

You found the how, a resolver, but the question is still if the warning isn't worse then the change...

> I opened https://github.com/numpy/numpy/pull/30194 (draft status) to see how involved the change with no warnings or deprecation would be.

Nice, let's use that to bring it up on the mailing list.  Also, should merge it after the 2.3 release which is a bit annoying to see if there is downstream fallout (e.g. pandas).

**Comment by WarrenWeckesser** [comment]
> You found the how, a resolver, ...

Ah, of course.

All your arguments for just changing it sound reasonable.

I'll send an email to the mailing list shortly.

**Comment by WarrenWeckesser** [comment]
> Also, should merge it after the 2.3 release ...

I assume you meant after 2.4, so it is in the main dev branch for long time before showing up in release 2.5.  Is that correct?
